zrml/swaps/src/lib.rs

```diff
@@ -1208,10 +1235,125 @@ mod pallet {
+        pub(crate) fn apply_to_cached_pools<F>(
+            pool_count: u32,
+            mutation: F,
+            max_weight_per_pool: Weight,
+        ) -> Weight
+        where
+            F: Fn(PoolId) -> Result<Weight, DispatchError>,
+        {
+            let mut total_weight = T::WeightInfo::apply_to_cached_pools_noop(pool_count);
+            for (index, (pool_id, _)) in PoolsCachedForArbitrage::<T>::drain().enumerate() {
+                // The mutation should never fail, but if it does, we just assume we
+                // consumed all the weight and rollback the pool.
+                let _ = with_transaction(|| match mutation(pool_id) {
+                    Err(err) => {
+                        log::warn!(
+                            "Arbitrage unexpectedly failed on pool {:?} with error: {:?}",
+                            pool_id,
+                            err,
+                        );
+                        total_weight = total_weight.saturating_add(max_weight_per_pool);
+                        TransactionOutcome::Rollback(err.into())
+                    }
+                    Ok(weight) => {
+                        total_weight = total_weight.saturating_add(weight);
+                        TransactionOutcome::Commit(Ok(()))
+                    }
+                });
+                if index.saturating_add(1) == (pool_count as usize) {
+                    break;
+                }
+            }
+            total_weight
+        }
+
+        // Execute arbitrage on a single pool.
+        pub(crate) fn execute_arbitrage(
+            pool_id: PoolId,
+            max_iterations: usize,
+        ) -> Result<Weight, DispatchError> {
+            let pool = Self::pool_by_id(pool_id)?;
+            let pool_account = Self::pool_account_id(pool_id);
+            let balances = pool
+                .assets
+                .iter()
+                .map(|a| (*a, T::AssetManager::free_balance(*a, &pool_account)))
+                .collect::<BTreeMap<_, _>>();
+            let asset_count = pool.assets.len() as u32;
+            let outcome_tokens = pool.assets.iter().filter(|a| **a != pool.base_asset);
+            let total_spot_price = pool.calc_total_spot_price(&balances)?;
+
+            if total_spot_price > BASE.saturating_add(ARBITRAGE_THRESHOLD) {
+                let (amount, iteration_count) =
+                    pool.calc_arbitrage_amount_mint_sell(&balances, max_iterations)?;
+                // We're faking a `buy_complete_sets` operation by transfering to the market prize
+                // pool.
+                T::AssetManager::transfer(
+                    pool.base_asset,
+                    &pool_account,
+                    &T::MarketCommons::market_account(pool.market_id),
+                    amount,
+                )?;
+                for t in outcome_tokens {
+                    T::AssetManager::deposit(*t, &pool_account, amount)?;
+                }
```

Comment:
> In theory, if the total spot price (sum of price of all assets denominated in ZTG) is higher than 1 ZTG, the outcome assets are too expensive (relative to the supply of ZTG in the prize pool).
> 
> So you increase the supply of outcome tokens to reduce indirectly the spot price of all assets equally (because `amount` is always equal for every asset).
> 
> And you decrease the supply of ZTG from the liquidity pool and increase it for the prize pool.

Yes, exactly. That's the essence of the arbitrage process. There are either too many outcome tokens, or too few. In the former case, we convert full sets to ZTG, in the latter case we convert some ZTG into outcome tokens.

Reply:
> In the [example](https://hackmd.io/@1ypDLjlbQ_e2Gp_1EW7kkg/BksyTQc-o#Example) Alice actually changed the spot price of A and B with a simple swap action.
> 
> > Alice swaps in 10 ZTG for A tokens
> 
> Why does Alice change the spot price of token B with that action from 0.5 to 0.55?
> 
> Through your transfer and deposit action I can not directly see, that the spot price of the assets changes (and your adjustments made therefore a difference in the result of the total spot price after the arbitrage). But I think this is the responsibility of the pool to recalculate the spot prices of the assets. I think I need to understand `calc_spot_price` for this first. But maybe you can give me a quick explanation for my question above.

Quantitatively, it's hard to say exactly why the price of B/ZTG goes to 0.55 and not 0.6, of course. It boils down to the spot price formula: `spot_price = (w_i / w_0) * (b_0 / b_i)`.

Qualitatively, the reason is this: When Alice buys A from the pool, she takes A out and adds ZTG to the pool. Removing A from the pool has no influence on the price of B/ZTG, but adding ZTG does. That means that each outcome token is now values against 60 ZTG instead of 50 ZTG, which makes each of them more valuable. You can see this in the formul above: The weights stay the same, so does `b_i` (the balance of the outcome token), but `b_0` increases, resulting in a higher spot price.